themes/default/content/registry/packages/aws/how-to-guides/bucketv2-migration.md

```diff
@@ -0,0 +1,2423 @@
+                {
+                    ["Null"] = new Dictionary<string, object?>
+                    {
+                        ["s3:x-amz-server-side-encryption-aws-kms-key-id"] = "true",
+                    },
+                },
+            },
+        })),
+    });
+
+});
+
+
+```
+
+{{% /choosable %}}
+
+{{% choosable language java %}}
+
+```java
+package generated_program;
+
+import com.pulumi.Context;
+import com.pulumi.Pulumi;
+import com.pulumi.core.Output;
+import com.pulumi.aws.s3.BucketV2;
+import com.pulumi.aws.s3.BucketPolicy;
+import com.pulumi.aws.s3.BucketPolicyArgs;
+import static com.pulumi.codegen.internal.Serialization.*;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class App {
+    public static void main(String[] args) {
+        Pulumi.run(App::stack);
+    }
+
+    public static void stack(Context ctx) {
+        var myBucket = new BucketV2("myBucket");
+
+        var myBucketPolicy = new BucketPolicy("myBucketPolicy", BucketPolicyArgs.builder()
+            .bucket("my-bucket-26224916")
+            .policy(myBucket.bucket().applyValue(bucket -> serializeJson(
+                jsonObject(
+                    jsonProperty("Version", "2012-10-17"),
+                    jsonProperty("Id", "PutObjPolicy"),
+                    jsonProperty("Statement", jsonObject(
+                        jsonProperty("Sid", "DenyObjectsThatAreNotSSEKMS"),
+                        jsonProperty("Principal", "*"),
+                        jsonProperty("Effect", "Deny"),
+                        jsonProperty("Action", "s3:PutObject"),
+                        jsonProperty("Resource", String.format("arn:aws:s3:::%s/*", bucket)),
+                        jsonProperty("Condition", jsonObject(
+                            jsonProperty("Null", jsonObject(
+                                jsonProperty("s3:x-amz-server-side-encryption-aws-kms-key-id", "true")
+                            ))
+                        ))
+                    ))
+                ))))
+            .build());
+
+    }
+}
+
+```
+
+{{% /choosable %}}
+
+{{% choosable language yaml %}}
+
+```yaml
+name: example
+runtime: yaml
+resources:
+  my-bucket:
+    type: aws:s3:BucketV2
+  my-bucket-policy:
+    type: aws:s3:BucketPolicy
+    properties:
+      bucket: "my-bucket-26224916"
+      policy:
+        fn::toJSON:
+          Version: 2012-10-17
+          Id: PutObjPolicy
+          Statement:
+            Sid: DenyObjectsThatAreNotSSEKMS
+            Principal: "*"
+            Effect: Deny
+            Action: s3:PutObject
+            Resource: "arn:aws:s3:::${my-bucket.bucket}/*"
+            Condition:
+              "Null":
+                s3:x-amz-server-side-encryption-aws-kms-key-id: "true"
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+
+
+
+
+As a bonus, the policy can now more easily refer to the concrete name of the bucket.
```

Comment:
It might be worth explaining the benefit of not having cyclic deps in a bit more detail in the beginning. Gives some more insights into why this is a good thing for users.

Reply:
It's a difficult sell. I tried adding a Context section at the bottom, but not sure it belongs, perhaps this sentence does not belong either. Primary motivation for users to do this is to get onto a supported version because they have to.

I've checked the style of https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade#s3-bucket-refactor FWIW it's very technical to-the-point and does not explain any design considerations as to why.